### PR TITLE
ENH: TST: add a threadpoolctl hook to limit OpenBLAS parallelism 

### LINKED ISSUE #14425: Running tests in parallel is not any faster than without pytest-xdist anymore
I have no idea what is going on here, but no matter how I run the tests, I'm not getting them to run any faster in parallel. These commands all give me the same timings (about 21 sec):
- `python -c "import scipy; scipy.test()"`
- `python -c "import scipy; scipy.test(parallel=4)"`
- `pytest --pyargs scipy.linalg`
- `pytest -n 4 --pyargs scipy.linalg`

And the same if I try it on other submodules or on scipy as a whole. I even uninstalled `pytest-xdist` to make sure - no difference whatsoever.

This used to work after gh-10172 two years ago. Does this still work for anyone else?

### PR DESCRIPTION
**Pull request by rgommers**
Closes gh-14425
    
This is taken over with minor modifications from scikit-learn. The most dramatic effect of this is for `scipy.linalg`, its test suite runs in ~30 sec on a 12-core machine without this change, and in ~5 sec with it when using `linalg.test(parallel=12)`. The full test suite is also sped up significantly.

### CODE REVIEW

File: scipy/conftest.py
Comment (peterbell10) [inline scipy/conftest.py:45]: `threadpoolctl` doesn't work well with `-OO`, see  joblib/threadpoolctl#87. In the mean time,

```suggestion
    except ImportError, AttributeError:
```

File: scipy/conftest.py
Comment (peterbell10) [inline scipy/conftest.py:63]: This is broken for old versions of OpenBLAS, see joblib/threadpoolctl#88.

**Comment by andyfaff** [comment]
If I understand correctly the issue is that there's oversubscription of the parallelisation: `pytest-xdist` is wanting to run tests in parallel by starting new processes, but `OpenBlas` is wanting to parallelise it's BLAS by using `omp`?

If that's correct, then the PR looks ok, but it seems @peterbell10 has more knowledge of `OpenBlas` and `threadpoolctl`, so I'll defer to him.

There's another takeaway point from this - how is normal user operation affected by parallelisation in `OpenBlas`? Say a user has their own parallelisation scheme (`mpi4py`/`multiprocessing`/pthread in an extension, and each of those threads/processes uses parallel `linalg`); in those instances they would need fine-grained control over what is allowed to parallelise otherwise you end up with oversubscription.

1. Does `OpenBlas` automatically use `omp` out of the box?
2. Is there a clear `scipy` API for controlling how the version of `OpenBlas` bundled by `scipy` uses parallelisation?
3. Same as 2, but for other libraries.

File: scipy/conftest.py
Comment (rgommers) [reply scipy/conftest.py:45]: Thanks for fixing things upstream Peter!

**Comment by rgommers** [comment]
> `pytest-xdist` is wanting to run tests in parallel by starting new processes, but `OpenBlas` is wanting to parallelise it's BLAS by using `omp`?

Yes that's correct. Additional complication is that OpenBLAS can be built with either its own threadpool or with OpenMP support. I believe pretty much everyone packages it with OpenMP though, and it's `threadpoolctl`'s job to deal with both situations.

> There's another takeaway point from this - how is normal user operation affected by parallelisation in `OpenBlas`? Say a user has their own parallelisation scheme (`mpi4py`/`multiprocessing`/pthread in an extension, and each of those threads/processes uses parallel `linalg`); in those instances they would need fine-grained control over what is allowed to parallelise otherwise you end up with oversubscription.

Exactly, that's already the case today. If you have linalg-heavy code and you want to use `multiprocessing` or Dask/Ray type packages, then you already need to manually deal with this oversubscription issue. A main reason we do not want to use OpenMP and have a `workers=` keyword to make SciPy parallelism opt-in is exactly this issue. Getting it right is nontrivial.

> 1. Does `OpenBlas` automatically use `omp` out of the box?

yes

> 2. Is there a clear `scipy` API for controlling how the version of `OpenBlas` bundled by `scipy` uses parallelisation?

No, this is what `threadpoolctl` is for. We had this discussion in NumPy, and that was the outcome: recommended to use `threadpoolctl`.

> 3. Same as 2, but for other libraries.

Same for scikit-learn, it has a dependency on `joblib` and `threadpoolctl` and its `n_jobs` keyword which is similar to our `workers` keyword.

File: scipy/conftest.py
Comment (peterbell10) [inline scipy/conftest.py:45]: Oops, my bad
```suggestion
    except (ImportError, AttributeError):
```

File: scipy/conftest.py
Comment (jjerphan) [inline scipy/conftest.py:61]: `os.cpu_count` returns the number of the system CPUs.
In some context, the process might be bound to only some of them.

For instance, you can see it in action using `taskset(1)`:
```sh
$ taskset -c 0-1 python -c "import os; print(os.cpu_count())"
8
```

One can use `os.sched_getaffinity`, which return the list of the usable CPUs.
```sh
$ taskset -c 0-1 python -c "import os; print(len(os.sched_getaffinity(0)))"
2
```

Might this be relevant here?

Comment (rgommers) [reply scipy/conftest.py:61]: Good question. Sounds like it could be, in particular on HPC systems. I stripped this implementation down from scikit-learn, which uses `joblib.cpu_count` that in the end is: https://github.com/joblib/loky/blob/6245bb2f7c16e283ced3d2ff25a9c9cc439a8867/loky/backend/context.py#L110. From further down in that file, `sched_getaffinity` is checked for but not always available.

This is a pretty deep rabbit hole, so I'm inclined to leave this as is - solves the problem for 95%. If we want to solve it for ~100%, then I'd prefer to add an optional dependency on `joblib` / `loky` (or vendor it), rather than try and reimplement it.

Comment (ogrisel) [reply scipy/conftest.py:61]: For reference the actual implementation is there:

https://github.com/joblib/loky/blob/6245bb2f7c16e283ced3d2ff25a9c9cc439a8867/loky/backend/context.py#L110-L255

It takes both CPU affinity and CFS CPU time quotas into account to compute the rounded actual amount of CPU allowed for the user's processes. It also has an option to count only physical cores instead of logical cores (hyperthreads) when no such restrictions are in place. It is quite often the case that hyperthreads are pretty useless for OpenBLAS.

Comment (rgommers) [reply scipy/conftest.py:61]: Thanks @ogrisel!

> It is quite often the case that hyperthreads are pretty useless for OpenBLAS.

Yes, I've noticed that too. The performance is only marginally better typically, and sometimes it can even degrade.

**Comment by rgommers** [comment]
Great, pytest is crashing so badly that I can't even get a traceback out of Azure (output too long, browser tab unresponsive).

The problem seems to be explained by https://github.com/pytest-dev/pytest/blob/134d8f78d886d5a552bd4f56209af5554bef14a1/doc/en/how-to/failures.rst#warning-about-unraisable-exceptions-and-unhandled-thread-exceptions (the `AttributeError` cannot be handled, triggers `PytestUnraisableExceptionWarning`, which then gets turned into an error).

Because I'm not sure how stable that pytest behavior is, I think it's better to just have an unconditional `except Exception`, with an explanation for why. The next version of `threadpoolctl` has the fix, but that doesn't help us now.

File: scipy/conftest.py
Comment (peterbell10) [reply scipy/conftest.py:68]: @rgommers "old versions of OpenBLAS" includes versions used in some of the CI. You need another `try-except`.

Comment (rgommers) [reply scipy/conftest.py:68]: Thanks, adding.

**Comment by rgommers** [comment]
Yep, we even have a test to warn us we are using the wrong warnings filter:
```
[gw1] linux -- Python 3.7.5 /usr/bin/python3.7
scipy/_lib/tests/test_warnings.py:100: in test_warning_calls_filters
    "\n    ".join(bad_filters)))
E   AssertionError: warning ignore filter should not be used, instead, use
E   numpy.testing.suppress_warnings (in tests only);
E   found in:
E       conftest.py:46
        bad_filters = ['conftest.py:46']
        bad_stacklevels = ['__init__.py:146', 'ndimage/_ni_support.py:85', 'ndimage/_interpolation.py:603', 'ndimage/_interpolation.py:798', 'linalg/_matfuncs_inv_ssq.py:827', 'linalg/_matfuncs_inv_ssq.py:836', ...]
        warning_calls = (['conftest.py:46', 'optimize/optimize.py:1068', 'sparse/__init__.py:245'], ['__init__.py:146', 'ndimage/_ni_support.p...n.py:603', 'ndimage/_interpolation.py:798', 'linalg/_matfuncs_inv_ssq.py:827', 'linalg/_matfuncs_inv_ssq.py:836', ...])
- g
```

**Comment by rgommers** [comment]
Hmm now it's failing with:
```
AttributeError: module 'pytest' has no attribute 'PytestUnraisableExceptionWarning'
```
on Windows. Looks like it was only added fairly recently (Dec 2020, pytest 6.2): https://github.com/pytest-dev/pytest/pull/8055. On Windows we still pick up pytest 5.4.3.

I'll change it to simply `pytest.PytestWarning`, which `PytestUnraisableExceptionWarning` inherits from.

**Comment by rgommers** [comment]
This seems to work now! CI failures are unrelated. Rebased but didn't make changes, the `threadpoolctl` 3.0.0 release from October fixed the warnings problem.

**Comment by rgommers** [comment]
Rebased once more because CI logs disappeared. I plan to merge this before the 1.8.0 release branch creation unless there are more comments. @peterbell10 would you mind having another look at this?

**Review by larsoner — APPROVED** [review]
LGTM +1 for merge

File: scipy/conftest.py
Comment (tylerjereddy) [inline scipy/conftest.py:8]: nit: `import numpy.typing as npt` is also used in the NumPy official docs: https://numpy.org/devdocs/reference/typing.html#number-precision

We're unlikely to import that in `conftest.py` anyway, so I don't really care much I don't think. I've seen this in other code bases too--perhaps any confusion on the standard alias will clear up over time.

**Review by tylerjereddy — APPROVED** [review]
It has core developer approval and CI seems ok--I guess we can just ignore the line length warnings in the diff lint.

The generic `Exception` catching is a bit unfortunate, but seems quite helpful to cut down our test suite times.

**Comment by tylerjereddy** [comment]
thanks Ralf et al.

File: scipy/conftest.py
Comment (rgommers) [reply scipy/conftest.py:8]: I don't mind either way, this abbreviation or just `from numpy.testing import what_one_needs`

**Pull request merged**
